crates/sdk/src/codec.rs

```diff
@@ -0,0 +1,544 @@
+};
+use p3_fri::CommitPhaseProofStep;
+
+use super::{F, SC}; // BabyBearPoseidon2Config
+
+type Challenge = BinomialExtensionField<F, 4>;
+
+/// Codec version should change only when proof system or proof format changes.
+/// It does correspond to the main openvm version (which may change more frequently).
+const CODEC_VERSION: u32 = 1;
+
+/// Hardware and language independent encoding.
+/// Uses the Writer pattern for more efficient encoding without intermediate buffers.
+// @dev Trait just for implementation sanity
+pub trait Encode {
+    /// Writes the encoded representation of `self` to the given writer.
+    fn encode<W: Write>(&self, writer: &mut W) -> Result<()>;
+
+    /// Convenience method to encode into a Vec<u8>
+    fn encode_to_vec(&self) -> Result<Vec<u8>> {
+        let mut buffer = Vec::new();
+        self.encode(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+/// Hardware and language independent decoding.
+/// Uses the Reader pattern for efficient decoding.
+pub trait Decode: Sized {
+    /// Reads and decodes a value from the given reader.
+    fn decode<R: Read>(reader: &mut R) -> Result<Self>;
+}
+
+// General Note [jpw]: even though `SC` is a concrete type below, for typedefs that rely on associated types such as PcsProof<SC>,
+// Rust will prevent you from implementing Encode on the typedef, (I think) because if the SC trait changes, then the associated
+// type may change. In these cases, either impl Encode on the concrete type, or make a separate `encode_something` function.
+
+pub fn encode_proof_to_bytes(proof: &Proof<SC>) -> Result<Vec<u8>> {
+    let mut buffer = Vec::new();
+    encode_proof(proof, &mut buffer)?;
+    Ok(buffer)
+}
+
+pub fn decode_proof_from_bytes(bytes: &[u8]) -> Result<Proof<SC>> {
+    let mut reader = Cursor::new(bytes);
+    decode_proof(&mut reader)
+}
+
+// ==================== Encode implementation ====================
+
+// We need to know:
+// - Pcs is TwoAdicFriPcs
+// - Com<SC>: Into<[F; 8]>
+// For simplicity, we only implement for fixed `BabyBearPoseidon2Config`
+//
+/// Encode a proof using FRI as the PCS with `BabyBearPoseidon2Config`.
+/// The Merkle tree hashes have digest `[F; 8]`.
+pub fn encode_proof<W: Write>(proof: &Proof<SC>, writer: &mut W) -> Result<()> {
+    writer.write_all(&CODEC_VERSION.to_le_bytes())?;
+    // Encode commitments
+    encode_commitments(&proof.commitments.main_trace, writer)?;
+    encode_commitments(&proof.commitments.after_challenge, writer)?;
+    let quotient_commit: [F; DIGEST_SIZE] = proof.commitments.quotient.into();
+    quotient_commit.encode(writer)?;
+
+    // Encode OpeningProof
+    encode_opening_proof(&proof.opening, writer)?;
+
+    // Encode per_air data
+    encode_slice(&proof.per_air, writer)?;
+
+    writer.write_all(&[RapPhaseSeqKind::FriLogUp as u8])?;
+    // Encode logup witness
+    proof.rap_phase_seq_proof.encode(writer)?;
+
+    Ok(())
+}
+
+// Helper function to encode OpeningProof
```

Comment:
```suggestion
// Helper function to encode OpeningProof
// ```
// pub struct OpeningProof<PcsProof, Challenge> {
//    pub proof: PcsProof,
//    pub values: OpenedValues<Challenge>,
// }
// ```
```